### ISSUE TITLE
Waste of bandwidth on first synching of Anoncoin without the bootstrap.dat

### ISSUE THREAD

**Issue by Cryptoslave** [Anoncoin/oldrepo-backup#88]
When installing a freshly compiled anoncoin 0.9.6.9 without any %appdata%/anoncoin directory, on windows, it will start to synch the blockchain. What happen is instead of taking 400 MB for the blockchain it takes up to 3 GB and a lot of time. This does not happen if using the bootstrap when after synching with the bootstrap the client is shutdown and then reopened. But, if after the bootstrap the client is left to synch normally without closing, then again a waste of bandwidth happen with a lot of time and data. 

This strange behaviour on the FIRST synching also happen with linux AFAIK. 

This is a low priority because the bootstrap is easy to use and then the program can be closed manually or automatically by uncommenting #stopafterblockimport=1 in anoncoin.conf. On next launch of the program the issue disappear and the block synchronization continue without trouble. No clue on what cause this behaviour.

**Comment by Cryptoslave** [comment]
In fact it is OK, what happen on first synching is it start to:
- download in parallel the block headers from all connected peers (headers first synchronisation)
- download blocks from peers and verify them

So block headers in fact takes 1.5 GB to synch up to block 450000, this is what looks like a waste of bandwidth, particularly when the bootstrap file is less than 400 MB, but bootstrap.dat is linearized and well it looks normal. This operate in a sort of burst mode in bandwidth. What happen is between the headers request, some blocks are verified, and all the time new peers are searched for connection making.

When all block headers are downloaded, the blocks data will be downloaded in parallel from all peers, and most of the computing power will be used to verify them. We can notice a steady stream of data downloaded and the processor turns alot. This step use less total bandwidth and is quicker.

When using the bootstrap, the second step is achieved more quickly as less headers are needed to synch up to now, so it looks like less a waste of bandwidth but in fact all this works perfectly I would say. Indeed after say 1GB of data was downloaded, blocks will only be around 15000 but headers will be up to 350000+, once headers are up to now, the blocks number will increase but again, with some lags due to processor working alot and reporting the processed blocks by mininum 100 in the debug.log but in the graphical interface it can go by 50000 in one huge stream. Indeed it does not hang on but is working hard to verify the scrypt hashes and compute the SHA256 true hashes as GroundRod implemented to make more easy future multichain integration.

**Comment by Cryptoslave** [comment]
After verification with Bitcoin 0.11.0, from an empty %appdata%/Bitcoin directory:

All 392387 headers of Bitcoin were recovered in 1 min for a total bandwidth of 39 MB

vs

All 471110 headers of Anoncoin were recovered in 8.5 hours for a total bandwidth of 5 GB, from ten peers in parallel.

==> There really is an issue with this code!

Issue is open again and the priority is increased as this is a serious bug

**Comment by Cryptoslave** [comment]
After verification with Bitcoin 0.11.0, from an empty %appdata%/Bitcoin directory:

All 392387 headers of Bitcoin were recovered in 1 min for a total bandwidth of 39 MB

vs

All 471110 headers of Anoncoin were recovered in 8.5 hours for a total bandwidth of 5 GB, from ten peers in parallel.And then it only started to download blocks...

==> There really is an issue with this code!

Issue is open again and the priority is increased as this is a serious bug

**Comment by Cryptoslave** [comment]
https://github.com/bitcoin/bitcoin/pull/4468/commits 
This we have to code/review

**Comment by Cryptoslave** [comment]
The issue was found to be this:
In the beginning there is like one line of : more getheaders (10000) to end to peer=1
for each peer one such line occur every 2000 unit

So for instance:
2016-01-09 00:10:39 received: headers (162003 bytes) from peer=1, peeraddr=b7ziruwpk7g2e44xyomnc2nu5tx7bc2f2ai4dzi66uxm3bc3qttq.b32.i2p
2016-01-09 00:10:40 more getheaders (70000) to end to peer=1, peeraddr=b7ziruwpk7g2e44xyomnc2nu5tx7bc2f2ai4dzi66uxm3bc3qttq.b32.i2p (startheight:471117)
2016-01-09 00:10:40 sending: getheaders (933 bytes) to peer=1, peeraddr=b7ziruwpk7g2e44xyomnc2nu5tx7bc2f2ai4dzi66uxm3bc3qttq.b32.i2p
....
2016-01-09 00:10:47 received: headers (162003 bytes) from peer=1, peeraddr=b7ziruwpk7g2e44xyomnc2nu5tx7bc2f2ai4dzi66uxm3bc3qttq.b32.i2p
2016-01-09 00:10:48 more getheaders (72000) to end to peer=1, peeraddr=b7ziruwpk7g2e44xyomnc2nu5tx7bc2f2ai4dzi66uxm3bc3qttq.b32.i2p (startheight:471117)
2016-01-09 00:10:48 sending: getheaders (933 bytes) to peer=1, peeraddr=b7ziruwpk7g2e44xyomnc2nu5tx7bc2f2ai4dzi66uxm3bc3qttq.b32.i2p

At first it run correctly, the more peers there are the more bandwidth wil be generated as each peer is requested the headers.

The problem is after block 100000 each peer is requested to send the headers several times:

2016-01-09 00:22:22 more getheaders (100000) to end to peer=1, peeraddr=b7ziruwpk7g2e44xyomnc2nu5tx7bc2f2ai4dzi66uxm3bc3qttq.b32.i2p (startheight:471117)
2016-01-09 00:22:22 received: headers (162003 bytes) from peer=1, peeraddr=b7ziruwpk7g2e44xyomnc2nu5tx7bc2f2ai4dzi66uxm3bc3qttq.b32.i2p
2016-01-09 00:22:23 more getheaders (100000) to end to peer=1, peeraddr=b7ziruwpk7g2e44xyomnc2nu5tx7bc2f2ai4dzi66uxm3bc3qttq.b32.i2p (startheight:471117)
2016-01-09 00:22:23 sending: getheaders (933 bytes) to peer=1, peeraddr=b7ziruwpk7g2e44xyomnc2nu5tx7bc2f2ai4dzi66uxm3bc3qttq.b32.i2p

twice for 100000 and peer 1

but

2016-01-09 01:27:21 more getheaders (170000) to end to peer=1, peeraddr=b7ziruwpk7g2e44xyomnc2nu5tx7bc2f2ai4dzi66uxm3bc3qttq.b32.i2p (startheight:471117)
2016-01-09 01:28:00 more getheaders (170000) to end to peer=1, peeraddr=b7ziruwpk7g2e44xyomnc2nu5tx7bc2f2ai4dzi66uxm3bc3qttq.b32.i2p (startheight:471117)
2016-01-09 01:28:37 more getheaders (170000) to end to peer=1, peeraddr=b7ziruwpk7g2e44xyomnc2nu5tx7bc2f2ai4dzi66uxm3bc3qttq.b32.i2p (startheight:471117)
2016-01-09 01:29:00 more getheaders (170000) to end to peer=1, peeraddr=b7ziruwpk7g2e44xyomnc2nu5tx7bc2f2ai4dzi66uxm3bc3qttq.b32.i2p (startheight:471117)
2016-01-09 01:30:07 more getheaders (170000) to end to peer=1, peeraddr=b7ziruwpk7g2e44xyomnc2nu5tx7bc2f2ai4dzi66uxm3bc3qttq.b32.i2p (startheight:471117)
2016-01-09 01:30:10 more getheaders (170000) to end to peer=1, peeraddr=b7ziruwpk7g2e44xyomnc2nu5tx7bc2f2ai4dzi66uxm3bc3qttq.b32.i2p (startheight:471117)
2016-01-09 01:30:57 more getheaders (170000) to end to peer=1, peeraddr=b7ziruwpk7g2e44xyomnc2nu5tx7bc2f2ai4dzi66uxm3bc3qttq.b32.i2p (startheight:471117)

So seven times for block headers 170000 and peer 1!

Each request generate 162003 bytes of headers, so for block header 70000 it was 162kB but for block header 170000 it was 1134 kB.

A quick calculus with 8 peers from begining and 471000 blocks show normally maximum 235 \* 8 requests of 162 kB each = 304 MB for all headers from all 8 peers. In fact it would be less than this because all peers are not connected from begining and sometimes the headers are synched from latter, not from the begining. So for bitcoin it was less than 50 MB for all headers and 8 peers.

But in Anoncoin, because of the multiplication of not useful requests for each peer, it can easily go to 2 to 5 GB for the whole synching of all block headers!! This is hence a total waste of bandwidth and of time and shall be corrected before release.

**Comment by Cryptoslave** [comment]
Corrected: https://github.com/Anoncoin/anoncoin/commit/0e35c180cd08bcc77b1ee94c454aac2170a85b8d

Around 2 hours and 50 MB to download the headers on clearnet, and check them.

https://wiki.anoncoin.net/images/f/f3/IBD_getheaderfirst_start_block_sync.PNG

After about 5 h in total (2 h + 3 h to download the blocks from all peers in parallel and check  them) and 440 MB of total bandwidth ANC finished to sync!

https://wiki.anoncoin.net/images/1/1a/IBD_getheaderfirst_finish_block_sync.PNG

Using the bootstrap.dat it takes around 3 h to sync and use 100 MB of bandwidth + the 103 MB of the bootstrap.tar.xz.

On I2P it takes around 10 hours to sync from scratch.
